Assessment/Assessment/ThreadSafeQueue.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <queue>
+#include <condition_variable>
+#include <shared_mutex>
+
+namespace Parallel
+{
+
+template<typename T>
+class ThreadSafeQueue
+{
+public: // public interface
+  typedef typename std::queue<T>::value_type value_type;
+  typedef typename std::queue<T>::size_type  size_type;
+
+public:
+  ThreadSafeQueue(size_type maxSize = 1000);
+
+  void push(const value_type v);
+  void pop(value_type& v);
```

Comment:
Also you can consider this form of pop function.
```cpp
// return true if value is popped, false if function return due to timeout
bool pop(value_type& value, time_type timeout);
```
It is perfectly fine to leave it as it is, but if someone waiting on pop and you decide to stop the application, you should take care to push some dummy value into queue to wake your pop from waiting state. But pushing dummy value will be possible in asynchronous application. In sync mode it will hang forever.
In the second form you can pop with some timeout and it will return as soon as new value inserted or timeout expired.